source/extensions/clusters/dynamic_forward_proxy/cluster.cc

```diff
@@ -82,10 +101,30 @@ void Cluster::onDnsHostAddOrUpdate(
   }
 
   ENVOY_LOG(debug, "adding new dfproxy cluster host '{}'", host);
+  Upstream::ClusterInfoConstSharedPtr cluster_info_to_use;
+  if (createCustomTlsForHost()) {
+    // Create an override cluster configuration that automatically provides both SNI as well as
+    // SAN verification for the resolved host if the cluster has been configured with TLS.
+    // TODO(mattklein123): The fact that we are copying the cluster config, etc. is not very clean.
```

Comment:
Did you try to put additional SAN in TransportSocketOptions? There is already SNI override there, if that works the copy can be avoided, as well as per host transport socket factory.

Reply:
I see, interesting. I think this is probably possible if the SAN verification could be overridden in this way, but I think it would require some plumbing changes since all of the verification right now is within the context itself, and we would need a callback from there into the transport socket I think, right? Unless you have more concrete hints and think it's easy, can I flesh out the TODO and we can consider this in a follow up? This definitely works, though I agree it's a little hacky.

Reply:
Let me spend a little time looking at this, I think it might not be that hard and I agree is definitely way better. (If you have any immediate tips on what APIs to use lmk.)

Reply:
Hmm, I don't see a version of `SSL_CTX_set_cert_verify_callback` that operates on the SSL vs. the CTX, at least not in https://github.com/google/boringssl/blob/master/include/openssl/ssl.h. @lizan @PiotrSikora @davidben how would I do this if I want to provide custom SAN verification per-SSL without an overriden CTX?

Reply:
Maybe I could set app data on the SSL, and then refer to that from within the verification callback to see if there are extra SANs to check? Thoughts?

Reply:
> Maybe I could set app data on the SSL, and then refer to that from within the verification callback to see if there are extra SANs to check? Thoughts?

Yeah from my reading this is probably right plumbing.

Reply:
OK let me give this a shot.